### PR TITLE
Fix 77603 - Don't break the log line when it is more than 16K

### LINKED ISSUE #77603: Pod logs: long lines are corrupted when using timestamps=true
**What happened**:

Docker has a hardcoded 16K limit on log line length, which splits lines longer than that into multiple messages in the log file. `ReadLogs` in [logs.go](https://github.com/kubernetes/kubernetes/blob/255a73ffc64d1c4e0f9a378ebf8d0005edfe6962/pkg/kubelet/kuberuntime/logs/logs.go), however, doesn't account for this when `timestamps` is set to `true` and prints the timestamp in front of every entry in the log file instead of every logical line. This leads to random timestamps appearing in the middle of log lines, potentially breaking the applications that parse these logs.

**What you expected to happen**:

`ReadLogs` inserts timestamps once per logical line, so original log data is preserved. Alternatively, allow `/logs` API to request the logs in a sane format like JSON, which would allow merging the lines back and combining them with the timestamps more intelligently on the application side.

**How to reproduce it (as minimally and precisely as possible)**:

* Create `foo.yaml`:

      apiVersion: v1
      kind: Pod
      metadata:
        name: log-lines
      spec:
        containers:
        - name: main
          image: busybox
          command: ['sh', '-c', 'while sleep 1; do echo -n ___; for i in $(seq 1 18000); do echo -n A; done; echo ___; done']

 * Start the pod with `kubectl apply -f foo.yaml`
 * Fetch the logs with `kubectl logs log-lines --timestamps`, or using the API (`curl https://…/api/v1/namespaces/default/pods/log-lines/log?timestamps=true&container=main`)
 * Observe random timestamps inserted in the middle of the log lines.

**Environment**:
- Kubernetes version (use `kubectl version`):
```Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.1", GitCommit:"b7394102d6ef778017f2ca4046abbaa23b88c290", GitTreeState:"clean", BuildDate:"2019-04-19T22:12:47Z", GoVersion:"go1.12.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.5", GitCommit:"2166946f41b36dea2c4626f90a77706f426cdea2", GitTreeState:"clean", BuildDate:"2019-03-25T15:19:22Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
```

### PR DESCRIPTION
**Pull request by EricWvi**
#### What type of PR is this?
/kind bug
/sig network

#### What this PR does / why we need it:
When the log line's content size is more than 16K, "kubectl logs --timestamps" breaks the line as multiple lines .

It is not convenient for other tool to parse the log
#### Which issue(s) this PR fixes:

Fixes #77603

#### release note
```release-note
When container log's length exceeds 16K, 
Docker will split the log entry into two entries. 
One contains "\n", the other not, 
which may affect log processing tools.

This fix also takes --tail, --timestamps, -f into account.
```

### CODE REVIEW

**Comment by EricWvi** [comment]
I moved the edits into a new branch, making previous PR closed. The previous PR is here #104321

**Comment by EricWvi** [comment]
The conversation is now at  
> dashpole
> We did a quick review of this during triage. Does this only impact kubectl logs when specifying `--timestamp`. Also, the linked issue is already closed. That may need to be reopened and discussed if it wasn't already fixed. Does this only impact kubectl logs, or does this change how logs are written to disk?

and I replied with

> Yeah, I wanted to re-open the issue, but failed. Thanks for @pacoxu , now it is open.
> The problem has not been fixed so far and I reproduced it in the cluster with Kubernetes 1.21.
> 
> ```
> Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.3", GitCommit:"ca643a4d1f7bfe34773c74f79527be4afd95bf39", GitTreeState:"clean", BuildDate:"2021-07-15T21:04:39Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"linux/amd64"}
> Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.3", GitCommit:"ca643a4d1f7bfe34773c74f79527be4afd95bf39", GitTreeState:"clean", BuildDate:"2021-07-15T20:59:07Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"linux/amd64"
> ```
> 
> #### Pod Details
> * one api server developed by gin
> 
> #### Scene
> * I deliberately insert one record into the database, which is a little bit longer than 16K.
> * I make a request for a document referring to that record.
> * `k logs --tail=8 --prefix --timestamps dockerlog` (I highlight the prefix and timestamps)
>   image
> * ”should halt an entire program or server request“ is the end of the document. It is splitted from the last log.
> * `k logs --tail=9 --prefix --timestamps dockerlog`, the timestamp is inserted between two logs. Besides, as for one piece of log, it is splitted into two.
>   image
>   ...(tl;dr)
>   image

File: pkg/kubelet/kuberuntime/logs/logs_test.go
Comment (endocrimes) [inline pkg/kubelet/kuberuntime/logs/logs_test.go:120]:
```suggestion
			name: "using TailLines 7 should output all lines when the log has <= 6 lines",
```

File: pkg/kubelet/kuberuntime/logs/logs_test.go
Comment (endocrimes) [inline pkg/kubelet/kuberuntime/logs/logs_test.go:113]:
```suggestion
			name: "using TailLines 3 should output the last 3 lines",
```

File: pkg/kubelet/kuberuntime/logs/logs.go
Comment (endocrimes) [inline pkg/kubelet/kuberuntime/logs/logs.go:279]: Could you add explicit tests for this function please - Otherwise when behavior drifts further from the util package this will become awkward to maintain.

Relatedly is there a reason that we aren't adding a modified version in pkg/util/tail?, this was the only usage of the function in k/k that I can find:

```
[kubernetes(master)] $ rg FindTailLineStartIndex
pkg/kubelet/kuberuntime/logs/logs.go
298:    start, err := tail.FindTailLineStartIndex(f, opts.tail)

pkg/util/tail/tail.go
64:// FindTailLineStartIndex returns the start of last nth line.
69:func FindTailLineStartIndex(f io.ReadSeeker, n int64) (int64, error) {

pkg/util/tail/tail_test.go
44:             s, err := FindTailLineStartIndex(r, test.n)
[kubernetes(master)] $
```

**Review by endocrimes — CHANGES_REQUESTED** [review]

File: pkg/kubelet/kuberuntime/logs/logs.go
Comment (EricWvi) [reply pkg/kubelet/kuberuntime/logs/logs.go:279]: I didn't mean to put the modified version here. But I did. There are two reasons. 

First, I think code in util are abstract from specific use case and reusable in many cases. The modified version is dedicated to way too long log. I also discovered that the FindTailLineStartIndex is used once, exactly here. It is really out of my expectation. I think the function is very useful to get the last few lines of a file.

Second, I put the  modified version here because I'm not sure where it should be. It caters to almostly only this situation. I need further advice.

After reviewers' suggestions about the code arrangement, I willl add explicit tests for this function.

**Comment by dims** [comment]
this PR has been on hold for about 2 weeks. Can you please see how to unblock this by reviewing comments etc? or please close this out (Trying to declutter our open list of PRs)

**Comment by dims** [comment]
There is a cncf-cla:no label on this PR. Can you please see the bot comments above to fix the CLA? This PR cannot progress unless the bot is happy with the CLA situation

**Comment by EricWvi** [comment]
I've reissued the pr, just fixing the line break when it is longer than 16K and not assuring the consistence between the requested number of lines and the actual number of output. I think one more log line or less makes no difference and it is kind of overkill.

**Pull request closed without merging**
